providers/base/bin/check_software_raid.py

```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/env python3
+# This file is part of Checkbox.
+#
+# Copyright 2024 Canonical Ltd.
+# Written by:
+#   Stanley Huang <stanley.huang@canonical.com>
+#
+# Checkbox is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3,
+# as published by the Free Software Foundation.
+#
+# Checkbox is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Checkbox. If not, see <http://www.gnu.org/licenses/>.
+
+import re
+import sys
+import logging
+import argparse
+import shlex
+import subprocess
+from pathlib import Path
+from typing import Dict
+
+
+class RAIDStats(Dict):
+    device = str
+    mode = str
+
+
+def get_md_stat(filename: str = "/proc/mdstat") -> list:
+    """
+    Parse the information of RAID devices from /dev/mdstat
+
+    Returns:
+        raid_stats (list): the RAID information
+    """
+    pattern = r"^(md[0-9]+) : active ([a-z0-9]+) [a-z0-9 \[\]]+$"
+    raid_raw_data = ""
+    raid_stats = []
+
+    with Path(filename) as node:
+        raid_raw_data = node.read_text().strip("\n")
+
+    for tmp_data in raid_raw_data.split("\n"):
+        find_result = re.search(pattern, tmp_data)
+        if find_result:
+            raid_node = RAIDStats(
+                device=find_result.groups()[0],
+                mode=find_result.groups()[1]
+            )
+            raid_stats.append(raid_node)
+
+    return raid_stats
+
+
+def dump_raid_info(nodes: list) -> None:
+    """
+    Display detail information for all of RAID devices
+
+    Args:
+        nodes (list): the name of MD devices
+    """
+    for node in nodes:
+        subprocess.run(shlex.split("mdadm -D /dev/{}".format(node)))
```

Comment:
If you want to test this function, you can mock `subprocess`, call `dump_raid_info` with a list of 2 nodes, and check that `subprocess.run` was called twice, for instance.

Reply:
added a unit test for dump_raid_info function, thanks.